Mathlib/RingTheory/DedekindDomain/FiniteAdeleRing.lean

```diff
@@ -364,6 +370,86 @@ instance : Algebra (R_hat R K) (FiniteAdeleRing R K) where
   commutes' _ _ := mul_comm _ _
   smul_def' r x := rfl
 
+instance : CoeFun (FiniteAdeleRing R K)
+    (fun _ ↦ ∀ (v : HeightOneSpectrum R), adicCompletion K v) where
+  coe a v := a.1 v
+
+open scoped algebraMap in
+lemma exists_finiteIntegralAdele_iff (a : FiniteAdeleRing R K) : (∃ c : R_hat R K,
+    a = c) ↔ ∀ (v : HeightOneSpectrum R), a v ∈ adicCompletionIntegers K v :=
+  ⟨by rintro ⟨c, rfl⟩ v; exact (c v).2, fun h ↦ ⟨fun v ↦ ⟨a v, h v⟩, rfl⟩⟩
+
+section Topology
+
+open Classical nonZeroDivisors Multiplicative Additive IsDedekindDomain.HeightOneSpectrum
+
+open scoped algebraMap -- coercion from R to FiniteAdeleRing R K
+open scoped DiscreteValuation
+
+variable {R K} in
+lemma clear_denominator (a : FiniteAdeleRing R K) :
+    ∃ (b : R⁰) (c : R_hat R K), a * ((b : R) : FiniteAdeleRing R K) = c := by
+  let S := {v | a v ∉ adicCompletionIntegers K v}
+  choose b hb h using clear_local_denominator (R := R) (K := K)
+  let p := ∏ᶠ v ∈ S, b v (a v)
+  have hp : p ∈ R⁰ := finprod_mem_induction (· ∈ R⁰) (one_mem _) (fun _ _ => mul_mem) <|
+    fun _ _ ↦ hb _ _
+  use ⟨p, hp⟩
+  rw [exists_finiteIntegralAdele_iff]
+  intro v
+  by_cases hv : a v ∈ adicCompletionIntegers K v
+  · exact mul_mem hv <| coe_mem_adicCompletionIntegers _ _
+  · change v ∈ S at hv
```

Comment:
Are we missing a `rfl` lemma here?

Reply:
`S` is a local definition `let S := {v | a v ∉ adicCompletionIntegers K v}` and I'm just changing `a v ∉ adicCompletionIntegers K v` to `v ∈ S`. But `change` is a bit of a code smell so I rewrote the proof to avoid it.